lib/Dialect/Arith/Transforms/QuarterWideInt.cpp

```diff
@@ -0,0 +1,516 @@
+    Location loc = op->getLoc();
+
+    auto newTy =
+        getTypeConverter()->convertType<RankedTensorType>(op.getType());
+    if (!newTy)
+      return rewriter.notifyMatchFailure(
+          loc, llvm::formatv("unsupported type: {0}", op.getType()));
+
+    Type newElemTy = reduceInnermostDim(newTy);
+
+    auto [lhsElem0, lhsElem1, lhsElem2, lhsElem3] =
+        extractLastDimHalves(rewriter, loc, adaptor.getLhs());
+    auto [rhsElem0, rhsElem1, rhsElem2, rhsElem3] =
+        extractLastDimHalves(rewriter, loc, adaptor.getRhs());
+
+    // Create Constant
+    auto intAttr = rewriter.getIntegerAttr(newElemTy, maxIntWidth >> 1);
+    auto constantOp = rewriter.create<mlir::arith::ConstantOp>(loc, intAttr);
+
+    auto lowSum0 =
+        rewriter.create<mlir::arith::AddIOp>(loc, lhsElem0, rhsElem0);
+    auto lowSum1 =
+        rewriter.create<mlir::arith::AddIOp>(loc, lhsElem1, rhsElem1);
+    auto lowSum2 =
+        rewriter.create<mlir::arith::AddIOp>(loc, lhsElem2, rhsElem2);
+    auto lowSum3 =
+        rewriter.create<mlir::arith::AddIOp>(loc, lhsElem3, rhsElem3);
+
+    auto carry0 = rewriter.create<mlir::arith::ShRSIOp>(loc, lowSum0,
+                                                        constantOp.getResult());
+    auto carry1 = rewriter.create<mlir::arith::ShRSIOp>(loc, lowSum1,
+                                                        constantOp.getResult());
+    auto carry2 = rewriter.create<mlir::arith::ShRSIOp>(loc, lowSum2,
+                                                        constantOp.getResult());
+
+    auto high1 = rewriter.create<mlir::arith::AddIOp>(loc, lowSum1, carry0);
+    auto high2 = rewriter.create<mlir::arith::AddIOp>(loc, lowSum2, carry1);
+    auto high3 = rewriter.create<mlir::arith::AddIOp>(loc, lowSum3, carry2);
+
+    Value resultVec = constructResultTensor(rewriter, loc, newTy,
+                                            {lowSum0, high1, high2, high3});
+    rewriter.replaceOp(op, resultVec);
+    return success();
+  }
+};
+
+struct ConvertMulI final : OpConversionPattern<mlir::arith::MulIOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      mlir::arith::MulIOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    Location loc = op->getLoc();
+
+    auto newTy =
+        getTypeConverter()->convertType<RankedTensorType>(op.getType());
+    if (!newTy)
+      return rewriter.notifyMatchFailure(
+          loc, llvm::formatv("unsupported type: {0}", op.getType()));
+
+    auto [lhsElem0, lhsElem1, lhsElem2, lhsElem3] =
+        extractLastDimHalves(rewriter, loc, adaptor.getLhs());
+    auto [rhsElem0, rhsElem1, rhsElem2, rhsElem3] =
+        extractLastDimHalves(rewriter, loc, adaptor.getRhs());
+
+    // FIXME: Implement the real Karatsuba algorithm for 4x4 multiplication.
+    // First part of Karatsuba algorithm
+    auto z00 = rewriter.create<mlir::arith::MulIOp>(loc, lhsElem0, rhsElem0);
```

Comment:
nit: all these create calls would be marginally simpler and easier to read if you used an ImplicitLocOpBuilder and did a using declaration for the arith ops

```cpp
using ::mlir::arith::MulIOp;
using ::mlir::arith::AddIOp;

...

  // in the current func
  
  ImplicitLocOpBuilder b(loc, rewriter);
  
  ...
  
  auto z02 = b.create<MulIOp>(lhsElem1, rhsElem1);
  auto z01_p1 = b.create<AddIOp>(lhsElem0, lhsElem1);
...
```